Deliverable/pgvector query

kuraisle opened the pull request:
## Pull Request Contents

| |
|-|
✨ Feature

## Description

I've added integration with the postgres/pgvector database for retrieving concepts by embedding.

This was achieved by:
- Adding a SQLAlchemy model for an embeddings table
  - Reads the table name and vector size from the environment
- Adding a `query_vector()` function to the omop queries that uses this table and joins with the concept table
- Modifying the embeddings code
  - Adding a `PGVectorQuery` component. This eschews haystack's pgvector integration, which can only use a single table, limiting what we can do with it, and not using the existing database connection
  - Modifying the Embedding class
    - Removing some methods
    - Modifying the get_retriever method so it configures a `PGVectorQuery` and returns it

I've tested and it works. The tests I've added are for the python, but I haven't added tests for the database side

## Related Issues or other material

- [X] This PR relates to one or more issues, detailed below

- Related Issue #113 
- Closes #113 

## ✅ Added/updated tests?

- [X] This PR contains relevant tests
  - Or doesn't need to per the below explanation


- [X] I've completed all **actions** and **tasks** detailed in the PR Template

kuraisle committed:
Update omop_models.py

Added sqlalchemy model for an embeddings table

kuraisle committed:
Added pgvector sqlalchemy

kuraisle committed:
Added pgvector integration

The RAG pipeline now uses a connection to an OMOP database with embeddings.

kuraisle committed:
Added testing for python parts

Not sure how best to test for the database interactions

kuraisle committed:
docs

bp1183 commented on lettuce/components/embeddings.py:
```diff
@@ -99,6 +92,30 @@ class EmbeddingModel(BaseModel):
+       query = query_vector(query_vector = query_embedding, n = top_k) 
+       query_results = self._connection.execute(query).mappings().all()
+       return {"documents": [
+               Document(
```
If `id`, `res`, and `score` are `None`, is this OK? Or should there be default values

kuraisle replied:
I've added an exception handler here. If these are `None`, that's a database issue, so failure is preferable to supplying defaults

bp1183 replied:
The only issue is `res.get("id")` will return `None` if no `id` key is present, so a `KeyError` will not be raised. I think should just be 
```
return {"documents": [
    Document(
        id=res["id"],
        content=res["content"],
        score=res.get["score"],
        ) for res in query_results]
}
```

kuraisle replied:
There was a reason I had it like this - something to do with what SQLAlchemy was returning. Whatever that reason was, it's out of date, and your suggestion works just fine. Updated!

bp1183 commented:
Overall I think LGTM!

kuraisle committed:
exception handling

kuraisle committed:
Merge branch 'main' into deliverable/pgvector-query

kuraisle committed:
Update check.run-tests.yml

The RAG pipelines require the DB_VECTABLE and DB_VECSIZE environment variables to be set so tests can pass

kuraisle commented on lettuce/omop/omop_queries.py:
```diff
@@ -174,3 +175,15 @@ def query_related_by_name(
 
 def query_related_by_id() -> Select: ...
+
+def query_vector(query_vector, n: int = 5) -> Select:
```
done, it's List[float]

kuraisle commented on lettuce/omop/omop_queries.py:
```diff
@@ -174,3 +175,15 @@ def query_related_by_name(
+        select(
+            Concept.concept_id.label("id"),
+            Concept.concept_name.label("content"),
+            Embedding.embedding.cosine_distance(query_vector).label("score"),
```
So I don't think that's the responsibility of this, it creates the query but doesn't execute. I've added an exception handler on Embeddings.get_retriever that checks and gives a (hopefully) informative error if the model vector dimensions don't match the DB_VECSIZE parameter

kuraisle committed:
Merge branch 'main' into deliverable/pgvector-query

kuraisle committed:
Update embeddings.py

kuraisle committed:
Merge branch 'main' into deliverable/pgvector-query
